drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,627 @@
+#define PINNACLE_REG_FEED_CONFIG3     (0x06) /* R/W */
+#define PINNACLE_REG_CAL_CONFIG1      (0x07) /* R/W */
+#define PINNACLE_REG_PS2_AUX_CONTROL  (0x08) /* R/W */
+#define PINNACLE_REG_SAMPLE_RATE      (0x09) /* R/W */
+#define PINNACLE_REG_Z_IDLE           (0x0A) /* R/W */
+#define PINNACLE_REG_Z_SCALER         (0x0B) /* R/W */
+#define PINNACLE_REG_SLEEP_INTERVAL   (0x0C) /* R/W */
+#define PINNACLE_REG_SLEEP_TIMER      (0x0D) /* R/W */
+#define PINNACLE_REG_EMI_THRESHOLD    (0x0E) /* R/W */
+#define PINNACLE_REG_PACKET_BYTE0     (0x12) /* R */
+#define PINNACLE_REG_PACKET_BYTE1     (0x13) /* R */
+#define PINNACLE_REG_PACKET_BYTE2     (0x14) /* R */
+#define PINNACLE_REG_PACKET_BYTE3     (0x15) /* R */
+#define PINNACLE_REG_PACKET_BYTE4     (0x16) /* R */
+#define PINNACLE_REG_PACKET_BYTE5     (0x17) /* R */
+#define PINNACLE_REG_GPIO_A_CTRL      (0x18) /* R/W */
+#define PINNACLE_REG_GPIO_A_DATA      (0x19) /* R/W */
+#define PINNACLE_REG_GPIO_B_CTRL_DATA (0x1A) /* R/W */
+#define PINNACLE_REG_ERA_VALUE        (0x1B) /* R/W */
+#define PINNACLE_REG_ERA_ADDR_HIGH    (0x1C) /* R/W */
+#define PINNACLE_REG_ERA_ADDR_LOW     (0x1D) /* R/W */
+#define PINNACLE_REG_ERA_ADDR_CTRL    (0x1E) /* R/W */
+#define PINNACLE_REG_PRODUCT_ID       (0x1F) /* R */
+
+/* Firmware ASIC ID value */
+#define PINNACLE_FIRMWARE_ID (0x07)
+
+/* Status1 definition */
+#define PINNACLE_STATUS1_SW_DR BIT(2)
+#define PINNACLE_STATUS1_SW_CC BIT(3)
+
+/* SysConfig1 definition */
+#define PINNACLE_SYS_CONFIG1_RESET          BIT(0)
+#define PINNACLE_SYS_CONFIG1_SHUTDOWN       BIT(1)
+#define PINNACLE_SYS_CONFIG1_LOW_POWER_MODE BIT(2)
+
+/* FeedConfig1 definition */
+#define PINNACLE_FEED_CONFIG1_FEED_ENABLE        BIT(0)
+#define PINNACLE_FEED_CONFIG1_DATA_MODE_ABSOLUTE BIT(1)
+#define PINNACLE_FEED_CONFIG1_FILTER_DISABLE     BIT(2)
+#define PINNACLE_FEED_CONFIG1_X_DISABLE          BIT(3)
+#define PINNACLE_FEED_CONFIG1_Y_DISABLE          BIT(4)
+/* X max to 0 */
+#define PINNACLE_FEED_CONFIG1_X_DATA_INVERT      BIT(6)
+/* Y max to 0 */
+#define PINNACLE_FEED_CONFIG1_Y_DATA_INVERT      BIT(7)
+
+/* FeedConfig2 definition */
+#define PINNACLE_FEED_CONFIG2_INTELLIMOUSE_ENABLE   BIT(0)
+#define PINNACLE_FEED_CONFIG2_ALL_TAPS_DISABLE      BIT(1)
+#define PINNACLE_FEED_CONFIG2_SECONDARY_TAP_DISABLE BIT(2)
+#define PINNACLE_FEED_CONFIG2_SCROLL_DISABLE        BIT(3)
+#define PINNACLE_FEED_CONFIG2_GLIDE_EXTEND_DISABLE  BIT(4)
+/* 90 degrees rotation */
+#define PINNACLE_FEED_CONFIG2_SWAP_X_AND_Y          BIT(7)
+
+/* Special definitions */
+#define PINNACLE_FB (0xFB) /* Filler byte */
+#define PINNACLE_FC (0xFC) /* Auto-increment byte */
+
+/* Read and write masks */
+#define PINNACLE_READ_MSK  (0xA0)
+#define PINNACLE_WRITE_MSK (0x80)
+
+/* Read and write register addresses */
+#define PINNACLE_READ_REG(addr)  (PINNACLE_READ_MSK | addr)
+#define PINNACLE_WRITE_REG(addr) (PINNACLE_WRITE_MSK | addr)
+
+struct pinnacle_active_range {
```

Comment:
one more general structure comment: I would recommend folding all these extra structures back into the main config one and then just passing `dev` as an argument to all functions, that way any driver function has access the whole state of the device, there's really not any performance or stability gain from segmenting the state into multiple structures and if you end up adding more configurations down the road you'll have to change the function signatures to get the rest of the data anyway

Reply:
Yeah that's understandable, I just found myself hacking drivers with this pattern in the past, missing the data I needed and reworking those to just get the `dev` argument instead, it's sort of a `self` pointer for the Zephyr driver model, at that point I think it's simpler to just pass that from the start everywhere you need any config or data. Thanks for doing the change.